Plugins/Detray/src/DetraySurfaceGridsConverter.cpp

```diff
@@ -0,0 +1,200 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Plugins/Detray/DetraySurfaceGridsConverter.hpp"
+
+#include "Acts/Detector/Detector.hpp"
+#include "Acts/Plugins/Detray/DetrayConversionUtils.hpp"
+#include "Acts/Plugins/Json/DetrayJsonHelper.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/Utilities/BinningType.hpp"
+#include "Acts/Utilities/GridAxisGenerators.hpp"
+
+#include <stdexcept>
+
+namespace Acts {
+
+/// DetraySurfaceGridsConverter converts surface grids from acts to detray
+/// format
+
+// convertAxis
+detray::io::axis_payload Acts::DetraySurfaceGridsConverter::convertAxis(
+    const Acts::IAxis& ia) {
+  detray::io::axis_payload axis_pd;
+  axis_pd.bounds = ia.getBoundaryType() == Acts::AxisBoundaryType::Bound
+                       ? detray::axis::bounds::e_closed
+                       : detray::axis::bounds::e_circular;
+  axis_pd.binning = ia.isEquidistant() ? detray::axis::binning::e_regular
+                                       : detray::axis::binning::e_irregular;
+  axis_pd.bins = ia.getNBins();
+  if (ia.isEquidistant()) {
+    axis_pd.edges = {ia.getBinEdges().front(), ia.getBinEdges().back()};
+  } else {
+    axis_pd.edges = ia.getBinEdges();
+  }
+
+  return axis_pd;
+}
+
+// convertGrid
+template <typename grid_type>
+detray::io::grid_payload<std::size_t, detray::io::accel_id>
+Acts::DetraySurfaceGridsConverter::convertGrid(const grid_type& grid,
+                                               bool swapAxis) {
+  // Get the grid axes & potentially swap them
+  detray::io::grid_payload<std::size_t, detray::io::accel_id> grid_pd;
+
+  auto axes = grid.axes();
+  if (swapAxis && grid_type::DIM == 2u) {
+    // std::cout<<"swap axes"<<std::endl;
+    std::swap(axes[0u], axes[1u]);
+  }
+
+  // Fill the axes in the order they are
+  for (unsigned int ia = 0u; ia < grid_type::DIM; ++ia) {
+    detray::io::axis_payload axis_pd = convertAxis(*axes[ia]);
+    axis_pd.label = static_cast<detray::axis::label>(ia);
+    grid_pd.axes.push_back(axis_pd);  // push axis to axes
+  }
+
+  // 1D connections
+  if constexpr (grid_type::DIM == 1u) {
+    for (unsigned int ib0 = 1u; ib0 <= axes[0u]->getNBins(); ++ib0) {
+      // Lookup bin
+      typename grid_type::index_t lbin;
+      detray::io::grid_bin_payload<std::size_t> grid_bin_pd;
+
+      lbin[0u] = ib0;
+      grid_bin_pd.content = grid.atLocalBins(lbin);
+      // Corrected bin for detray
+      lbin[0u] = ib0 - 1u;
+      grid_bin_pd.loc_index =
+          std::vector<unsigned int>(lbin.begin(), lbin.end());
+      grid_pd.bins.push_back(grid_bin_pd);
+    }
+  }
+
+  // 2D connections
+  if constexpr (grid_type::DIM == 2u) {
+    for (unsigned int ib0 = 1u; ib0 <= axes[0u]->getNBins(); ++ib0) {
+      for (unsigned int ib1 = 1u; ib1 <= axes[1u]->getNBins(); ++ib1) {
+        typename grid_type::index_t lbin;
+        // Lookup bin - respect swap (if it happened) for the lookup
+        lbin[0u] = swapAxis ? ib1 : ib0;
```

Comment:
Were these not already swapped in line 54?

Reply:
I think that's correct.